### PR TITLE
chore: use a custom presubmit_tests.sh script

### PR DESCRIPTION
**Pull request by lance**
The Knative test infrastructure runs presubmit tests which are defined in vendor/knative.dev/hack/presubmit-tests.sh. This has been running in test and release infrastructure. It is possible to customize the behavior of this script by defining a handful of functions. The function definitions provided here ensure that we are running tasks that are already being run with the default (e.g. verify-codgen.sh) but make use of our existing build and test targets.

(There also seems to be a stray version of this that I committed a couple of years ago. That has been removed as it is in the wrong location for test-infra and was never used anyway.)

Relates to: https://github.com/knative/func/issues/1333

EDIT: This PR focuses entirely on unit tests. A follow-up PR addressing integration tests will come next.

See: https://github.com/knative/infra/pull/115 (configures prow for func unit tests)
See: https://github.com/knative/infra/pull/104 (reverted, but this includes changes required for integration tests)

/kind chore

### CODE REVIEW

File: hack/presubmit-tests.sh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  header "Running integration tests"
+  make test-integration || failed=1
+
+  if (( failed )); then
+    results_banner "Integration tests failed"
+    exit ${failed}
+  fi
+}
+
+main "$@"
```
Comment (matejvasek) [inline hack/presubmit-tests.sh:82]: nit: missing newline

File: hack/presubmit-tests.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  if (( failed )); then
+    results_banner "Template tests failed"
+    exit ${failed}
+  fi
+}
+
+function integration_tests() {
+  local failed=0
+  header "Running integration tests"
+  make test-integration || failed=1
```
Comment (matejvasek) [inline hack/presubmit-tests.sh:74]: Where will this run? Integration tests requires some set up (e.g. docker daemon, cluster, registry).

**Comment by matejvasek** [comment]
Where is `presubmit-tests.sh` described? I don't see it in client/serving/eventing.

**Comment by lance** [comment]
Not sure why codecov has failed - I have not reduced the total tests. I will look into it, but I think maybe we shouldn't let that hold this up.

**Comment by lance** [comment]
> Not sure why codecov has failed - I have not reduced the total tests. I will look into it, but I think maybe we shouldn't let that hold this up.

Rebased and force pushed to trigger actions/jobs. We'll see if that changes codecov reporting.

**Comment by lkingland** [comment]
The code coverage test doesn't seem to behave deterministically, and we appear to add an exception to omit style-checking the `vendor` directory
/lgtm

**Review by lkingland — APPROVED** [review]

**Pull request merged**